### PR TITLE
broadcast is in to support graceful shutdown #8

### PR DESCRIPTION
**Pull request by victusfate**
I needed broadcast support in to properly do graceful shutdown in sp0n-sockets, which is next.

The core of this branch is an iteration over connected client sockets and sending a fixed payload to them all.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -162,10 +162,9 @@ class WebSocketJSONServer extends EventEmitter {
 
   }
 
-  // payload format: { type: 'messageType', data: $someJSON }
-  sendJson(ws,payload) {
+  send(ws,payload) {
```
Comment (victusfate) [inline index.js:165]: I thought I would need this function so I exposed it.
Now you can send binary, cars, whatever and not just json.

File: index.js [hunk 2]
```diff
@@ -176,6 +175,35 @@ class WebSocketJSONServer extends EventEmitter {
     })
   }
 
+
+  // payload format: { type: 'messageType', data: $someJSON }
+  sendJson(ws,payload) {
+    return this.send(ws,JSON.stringify(payload));
+  }
+
+
```
Comment (victusfate) [inline index.js:184]: broadcast is new but not hot. It just walks connections and sends them all a payload

File: test/test-app.js [hunk 3]
```diff
@@ -77,17 +77,28 @@ const wss             = new wsJson.WebSocketJSONServer({ server: server, verifyC
 // patchEmitter(wss)
 
 
+const idFromSocket = (ws) => {
+  return ws && ws.upgradeReq && ws.upgradeReq.decoded ? ws.upgradeReq.decoded.socketId : null;
+}
+
```
Comment (victusfate) [inline test/test-app.js:83]: idFromSocket grabs a decoded header property after client socket validation

File: test/test-appBroadcast.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+wss.on('messageJson', options => {
+  const ws        = options.ws;
+  const data      = options.data;
+  const socketId  = idFromSocket(ws);  
+  console.log({ action: 'wss.messageJson', data: data })
+
+  console.log({ action: 'wss.messageJson', socketId: socketId, data: data })
+  const oBroadcast = { type: data.type + '.broadcast', data: { from: socketId, date: Date.now() }, hash: data.hash};
+  wss.broadcastJson(oBroadcast)
```
Comment (victusfate) [inline test/test-appBroadcast.js:91]: new test with like 1 line different...

File: test/test-broadcastAndClients.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    console.log(`stderr: ${data}`);
+  });
+
+  client.on('close', (code) => {
+    console.log(`child process exited with code ${code}`);
+  });   
+}
+
+spawnClient();
+spawnClient();
```
Comment (victusfate) [inline test/test-broadcastAndClients.js:21]: this just kicks up a local wsJson server with built in auto broadcast replies for testing

File: test/test-client.js [hunk 6]
```diff
@@ -3,18 +3,19 @@
...
+const secret              = 'someSharedSecret'; 
+const port                = 3000;
+const sSocketUrl          = `ws://localhost:${port}`
 
 
 const createToken = (options) => {
   const sAction = 'createToken';
   console.log({ action: 'createToken' });
-  options = options ? options : { data: {} };
+  options = options ? options : { data: { socketId: uuid() } };
```
Comment (victusfate) [inline test/test-client.js:18]: now these test clients get an identity

File: index.js [hunk 7]
```diff
@@ -176,6 +175,35 @@ class WebSocketJSONServer extends EventEmitter {
...
+  broadcastJson(payload) {
+    const oJsonPayload = JSON.stringify(payload);
```
Comment (micahbrich) [inline index.js]: aha! an object that is actually a string! i think.

Comment (victusfate) [reply index.js]: OMG

Comment (victusfate) [reply index.js]: fixed!

**Pull request merged**